### ISSUE TITLE
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored

### ISSUE THREAD

**Issue by sobolevn** [typescript-eslint/typescript-eslint#1333]
**What code were you trying to parse?**

I am trying to work with this setup: https://github.com/wemake-services/wemake-vue-template/blob/master/template/.eslintrc.js

Here's how other eslint reusable packages from there are defined:
- https://github.com/wemake-services/wemake-frontend-styleguide/blob/master/packages/eslint-config-typescript/base.js

The problem might be in the amount of packages there.

**What did you expect to happen?**

It should work correctly.

**What actually happened?**

```
» npm run lint:js

> name@0.0.0 lint:js /Users/sobolev/Documents/github/wemake-vue-template/template
> eslint --cache --ext .js,.ts,.vue,.json --ignore-path .gitignore .

`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.11.0` |
| `TypeScript`                | `3.7.3` |
| `ESLint`                    | `6.7.2` |
| `node`                      | `10.16` |
| `npm`                       | `6.11` |

**Comment by bradzacher** [comment]
I don't know where that error message comes from.
We don't have that message anywhere in our codebase, and a quick search of the eslint codebase, it's not there either.

considering how many plugins and extensions there are, it's hard to tell what your config is even doing.
You can use the `--print-config` argument for eslint to show the fully resolved config.

**Comment by sobolevn** [comment]
Sure, here you go! https://gist.github.com/sobolevn/0e0e13ac80c5730a6a0c76609b6f046d

**Comment by bradzacher** [comment]
It's a problem with your `eslint-plugin-import` config.

https://github.com/benmosher/eslint-plugin-import/blob/813a116f93d160d349631ee5598fd387a38a9d1e/utils/parse.js#L53-L57

Something in your web of config dependencies is causing a problem with the parser. This is known working in other setups, so I don't know what's causing the problem in yours.

**Comment by chenshuai2144** [comment]
update doc：
https://github.com/typescript-eslint/typescript-eslint/blob/f335c504bcf75623d2d671e2e784b047e5e186b9/docs/getting-started/linting/FAQ.md

**Comment by sobolevn** [comment]
Turned out that disabling `'import/no-unused-modules': ['error', { 'unusedExports': true }],` worked.
Now this warning is gone.

I will open a new issue for `eslint-plugin-import`

**Comment by ljharb** [comment]
I’m not sure how this isn’t a bug in the parser - if parseForEslint doesn’t return a valid ast then it seems like the parser is broken.

Disabling rules as a perf recommendation doesn’t change that the rules should still *work* when enabled.

**Comment by bradzacher** [comment]
If there was a parser error when invoked by `eslint-plugin-import`, then it would show up as part of the normal lint run; `eslint-plugin-import` doesn't to anything special with its parse.

The fact that there is no parsing error reported by the user in the normal lint run suggests that there is a different issue, one likely with their config.

I guess there is a slim chance that `eslint-plugin-import` is parsing an npm dependency, which is throwing (which wouldn't happen as part of a normal lint run), but that's seems unlikely?

It's hard to tell because the plugin swallows the error, if there was one.

Didn't seen like it was worth the time to investigate further at the time; as it was the first and only report of this, and we recommend not using those rules anyways.

If more people are running into it, we can look at reopening and investigating further.

**Comment by bradzacher** [comment]
@sobolevn could you list the exact repro steps for your repo? 

Is it just clone the repo linked in the post (https://github.com/wemake-services/wemake-vue-template/), npm install, `npm run lint:js`?

**Comment by sobolevn** [comment]
@bradleyayers it would be easier with https://github.com/wemake-services/wemake-frontend-styleguide

Steps:
1. Clone the repo and run `yarn install`
2. Uncomment this line: https://github.com/wemake-services/wemake-frontend-styleguide/blob/master/packages/eslint-config-typescript/base.js#L71
3. Run `yarn lint:ts`

It will show the warning. Thanks for your help!

**Comment by bradzacher** [comment]
After doing the above, I replaced the swallowing try/catch here:
https://github.com/benmosher/eslint-plugin-import/blob/813a116f93d160d349631ee5598fd387a38a9d1e/utils/parse.js#L49-L51
With
```ts
} catch (e) {
  //
  console.error(e);
  console.error(parserOptions.filePath);
}
```

And got the following output when I ran `yarn lint:ts`:
```
{ index: 102,
  lineNumber: 7,
  column: 2,
  message: "Expected corresponding JSX closing tag for 'input'." }
'/Users/bradzacher/temp/wemake-frontend-styleguide-master/packages/eslint-config-vue/tests/fixtures/Correct.vue'
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
```
and
```
{ index: 137, lineNumber: 9, column: 6, message: "'}' expected." }
'/Users/bradzacher/temp/wemake-frontend-styleguide-master/packages/eslint-config-vue/tests/fixtures/Incorrect.vue'
`parseForESLint` from parser `@typescript-eslint/parser` is invalid and will just be ignored
```

Digging into it, `eslint-plugin-import` is attempting to gather exports for all the files in the project (including vue files), but because the eslint config is for typescript, not vue, this is why it fails to parse the vue files.

**Comment by ljharb** [comment]
That implies the user’s eslint config needs to exclude .vue flies from being parsed; their glob is probably over-eager.

A PR to eslint-plugin-import to log those error messages would be happily accepted.

**Comment by sobolevn** [comment]
@ljharb how can I make it work with `.vue` files as well? Because my `.vue` files are written in typescript and I do import logic in them. Without `.vue` files support `'import/no-unused-modules'` does not make any sense. 

Final config for Vue: https://github.com/wemake-services/wemake-vue-template/blob/master/template/.eslintrc.js

**Comment by armano2** [comment]
@sobolevn To make it work with vue files you have to add `"extraFileExtensions: [".vue"].` to your eslint config.
https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#extrafileextensions

**Comment by bradzacher** [comment]
In this case, nah it's not that - the vue files are vue.
The problem is having separate lint commands which don't properly include/exclude the files that they support.

Your typescript lint command shouldn't attempt to lint vue files, because that won't work.

Solutions are either:
- Add the vue eslint parser to your typescript config so it can handle vue files
- Explicitly exclude vue files from your typescript lint command.

**Comment by sobolevn** [comment]
That is strange. How do `.vue` files end up in my linting target when I explicitly define both `--config` and `--ext` options for `eslint`? Here: https://github.com/wemake-services/wemake-frontend-styleguide/blob/master/package.json#L9

```
"lint:ts": "eslint --cache --config ./packages/eslint-config-typescript/recommended.js --ext .ts --ignore-pattern incorrect.ts packages/eslint-config-typescript/tests/fixtures",
    
```

**Comment by RobTS** [comment]
I did the same as @bradzacher and for me, the files are as follows:

./node_modules/@react-native-community/async-storage/jest/async-storage-mock.js

./node_modules/react-native-keyboard-aware-scroll-view/lib/KeyboardAwareHOC.js

./node_modules/@react-native-community/async-storage/lib/AsyncStorage.js

./node_modules/@storybook/addons/src/index.js

Why is my node_modules folder even being parsed, shouldn't it be disabled by default?

**Comment by bradzacher** [comment]
eslint-plugin-import parses the files touched by _all_ import/require statements.
That includes files in node_modules. This parsing is separate to the standard eslint parsing cycle.

This is done so that it can collect information about the exports of the file, so that it can report if you're importing names/default correctly, etc.

This is the reason I suggest TS users don't use the rules which require a parser - because TS itself does this validation as part of the standard typecheck, so it's double handling.

https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#eslint-plugin-import

**Comment by kaiyoma** [comment]
We're hitting this error as well in our project, but I'm unclear how to debug this. Most of the comments at the end of this thread talk about Vue, but we don't use Vue. The error message isn't helpful at all, doesn't say what plugin is at fault, and doesn't say what file is triggering the error.

https://github.com/typescript-eslint/typescript-eslint/issues/1333#issuecomment-573353494 contains a hack to print out a bunch of helpful information in this scenario. Why don't you merge that so people can debug this themselves in the future?

**Comment by bradzacher** [comment]
Steps to help you figure out what's happening and what might be wrong:

1) Read through the docs on how to setup your project:
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/README.md

2) Read through the FAQ:
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md

3) Inspect your tsconfig - make sure that all the files you want to lint are in your `include`.

4) If all of that looks fine, then use `DEBUG=typescript-eslint:* yarn lint` to help inspect what the parser is doing.

> Why don't you merge that so people can debug this themselves in the future

It's open source! If you think it's valuable, then it's just a PR away.
